### ISSUE TITLE
Compiling in MSYS fails loudly.

### ISSUE THREAD

**Issue by IngwiePhoenix** [Midar/objfw#14]
Compiling under MSYS, using the mingw-w64 tools, causes pretty loud failures.

- `Sleep` is always "implicitly declared" and when running with threads enabled, at one point, a huge error will be thrown saying that the definitions are not matching.
```
Successfully compiled lookup.m (lib).
In file included from property.m:27:
..\threading.h:215:3: error: implicit declaration of function 'Sleep' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                Sleep(0);
                ^
1 error generated.
Failed to compile property.m (lib)!
../../buildsys.mk:491: die Regel für Ziel „property.lib.o“ scheiterte
make[5]: *** [property.lib.o] Fehler 1
../../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[4]: *** [all] Fehler 2
../buildsys.mk:129: die Regel für Ziel „runtime“ scheiterte
make[3]: *** [runtime] Fehler 2
../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[2]: *** [all] Fehler 2
buildsys.mk:129: die Regel für Ziel „src“ scheiterte
make[1]: *** [src] Fehler 2
buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2
```
I did edit `buildsys.mk` which then led me to the definition error (I removed `-Werror`)

- Running with threads disabled, does:
```
Successfully compiled windows_1252.m (lib).
OFDataArray.lib.o:(.data+0x928): multiple definition of `__objc_ivar_offset_OFDataArray._itemSize'
OFBigDataArray.lib.o:(.data[__objc_ivar_offset_OFDataArray._itemSize]+0x0): first defined here
OFDataArray.lib.o: In function `-[OFDataArray initWithContentsOfURL:]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFDataArray.m:214: multiple definition of `__objc_ivar_offset_OFDataArray._items'
OFBigDataArray.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFBigDataArray.m:29: first defined here
OFDataArray.lib.o: In function `-[OFDataArray initWithContentsOfURL:]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFDataArray.m:214: multiple definition of `__objc_ivar_offset_OFDataArray._capacity'
OFBigDataArray.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFBigDataArray.m:29: first defined here
OFDataArray.lib.o: In function `-[OFDataArray initWithContentsOfURL:]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFDataArray.m:233: multiple definition of `__objc_ivar_offset_OFDataArray._count'
OFBigDataArray.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFBigDataArray.m:29: first defined here
OFInflate64Stream.lib.o:(.data[__objc_ivar_offset_OFInflateStream._atEndOfStream]+0x0): multiple definition of `__objc_ivar_offset_OFInflateStream._atEndOfStream'
OFInflateStream.lib.o:(.data+0x6d0): first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._state'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:374: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._inLastBlock'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:374: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._stream'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._buffer'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._bufferIndex'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._bufferLength'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._bitIndex'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._context'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._slidingWindow'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:375: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._slidingWindowMask'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:373: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._slidingWindowIndex'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:375: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._savedBits'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:466: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._savedBitsLength'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:466: first defined here
OFInflate64Stream.lib.o: In function `+[OFInflate64Stream initialize]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/./OFInflateStream.m:256: multiple definition of `__objc_ivar_offset_OFInflateStream._byte'
OFInflateStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFInflateStream.m:463: first defined here
OFSHA224Or256Hash.lib.o:(.data+0x458): multiple definition of `__objc_ivar_offset_OFSHA224Or256Hash._state'
OFSHA224Hash.lib.o:(.data[__objc_ivar_offset_OFSHA224Or256Hash._state]+0x0): first defined here
OFSHA384Or512Hash.lib.o:(.data+0x458): multiple definition of `__objc_ivar_offset_OFSHA384Or512Hash._state'
OFSHA384Hash.lib.o:(.data[__objc_ivar_offset_OFSHA384Or512Hash._state]+0x0): first defined here
OFSortedList.lib.o:(.data[__objc_ivar_offset_OFList._lastListObject]+0x0): multiple definition of `__objc_ivar_offset_OFList._lastListObject'
OFList.lib.o:(.data+0x6e0): first defined here
OFStream.lib.o:(.data+0x6c0): multiple definition of `__objc_ivar_offset_OFStream._readBufferLength'
OFSeekableStream.lib.o:(.data[__objc_ivar_offset_OFStream._readBufferLength]+0x0): first defined here
OFStream.lib.o: In function `-[OFStream readBigEndianFloat]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFStream.m:242: multiple definition of `__objc_ivar_offset_OFStream._readBuffer'
OFSeekableStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFSeekableStream.m:27: first defined here
OFXMLElement.lib.o:(.data[__objc_ivar_offset_OFXMLAttribute._name]+0x0): multiple definition of `__objc_ivar_offset_OFXMLAttribute._name'
OFXMLAttribute.lib.o:(.data+0x6b8): first defined here
OFXMLElement.lib.o: In function `_references_to_categories_of_OFXMLElement':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFXMLElement.m:42: multiple definition of `__objc_ivar_offset_OFXMLAttribute._namespace'
OFXMLAttribute.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFXMLAttribute.m:124: first defined here
OFXMLElement+Serialization.lib.o:(.data[__objc_ivar_offset_OFXMLElement._name]+0x0): multiple definition of `__objc_ivar_offset_OFXMLElement._name'
OFXMLElement.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFXMLElement.m:211: first defined here
OFHTTPServer.lib.o:(.data[__objc_ivar_offset_OFHTTPResponse._statusCode]+0x0): multiple definition of `__objc_ivar_offset_OFHTTPResponse._statusCode'
OFHTTPResponse.lib.o:(.data+0x8c8): first defined here
OFHTTPServer.lib.o: In function `-[OFHTTPServerResponse initWithSocket:server:]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFHTTPServer.m:185: multiple definition of `__objc_ivar_offset_OFHTTPResponse._headers'
OFHTTPResponse.lib.o:(.data+0x880): first defined here
OFStreamSocket.lib.o:(.data[__objc_ivar_offset_OFStream._blocking]+0x0): multiple definition of `__objc_ivar_offset_OFStream._blocking'
OFStream.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src/OFStream.m:242: first defined here
OFTCPSocket.lib.o:(.data[__objc_ivar_offset_OFStreamSocket._socket]+0x0): multiple definition of `__objc_ivar_offset_OFStreamSocket._socket'
OFStreamSocket.lib.o:(.data+0x6c0): first defined here
OFArray_adjacentSubarray.lib.o:(.data[__objc_ivar_offset_OFArray_subarray._array]+0x0): multiple definition of `__objc_ivar_offset_OFArray_subarray._array'
OFArray_subarray.lib.o:(.data+0x628): first defined here
OFArray_adjacentSubarray.lib.o: In function `-[OFArray_adjacentSubarray objects]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFArray_adjacentSubarray.m:24: multiple definition of `__objc_ivar_offset_OFArray_subarray._range'
OFArray_subarray.lib.o:(.data+0x630): first defined here
OFString_UTF8.lib.o:(.data+0x6c0): multiple definition of `__objc_ivar_offset_OFString_UTF8._s'
OFMutableString_UTF8.lib.o:(.data[__objc_ivar_offset_OFString_UTF8._s]+0x0): first defined here
.exceptions_exceptions.lib.a.objs/OFReadOrWriteFailedException.lib.o:(.data+0x4f8): multiple definition of `__objc_ivar_offset_OFReadOrWriteFailedException._object'
.exceptions_exceptions.lib.a.objs/OFReadFailedException.lib.o:(.data[__objc_ivar_offset_OFReadOrWriteFailedException._object]+0x0): first defined here
.exceptions_exceptions.lib.a.objs/OFReadOrWriteFailedException.lib.o:(.data+0x500): multiple definition of `__objc_ivar_offset_OFReadOrWriteFailedException._requestedLength'
.exceptions_exceptions.lib.a.objs/OFReadFailedException.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src\exceptions/OFReadFailedException.m:23: first defined here
.exceptions_exceptions.lib.a.objs/OFReadOrWriteFailedException.lib.o:(.data+0x508): multiple definition of `__objc_ivar_offset_OFReadOrWriteFailedException._errNo'
.exceptions_exceptions.lib.a.objs/OFReadFailedException.lib.o:C:\Users\kingwersen\Downloads\objfw-0.8\src\exceptions/OFReadFailedException.m:23: first defined here
OFKernelEventObserver.lib.o: In function `+[OFKernelEventObserver alloc]':
C:\Users\kingwersen\Downloads\objfw-0.8\src/OFKernelEventObserver.m:92: undefined reference to `_OBJC_CLASS_OFKernelEventObserver_select'
OFKernelEventObserver.lib.o:(.rdata[__objc_class_ref_OFKernelEventObserver_select]+0x0): undefined reference to `__objc_class_name_OFKernelEventObserver_select'
clang.exe: error: linker command failed with exit code 1 (use -v to see invocation)
Failed to link libobjfw.so!
../buildsys.mk:238: die Regel für Ziel „libobjfw.so“ scheiterte
make[3]: *** [libobjfw.so] Fehler 1
../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[2]: *** [all] Fehler 2
buildsys.mk:129: die Regel für Ziel „src“ scheiterte
make[1]: *** [src] Fehler 2
buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2
```

My setup:
- gcc: 5.2.0
- clang: 3.6.2

Could you possibly look into this? Thanks!

**Comment by IngwiePhoenix** [comment]
Something I should probably mention:

```
$  gcc -dM -E - < /dev/null | grep -i mingw
#define __MINGW32__ 1
#define __MINGW64__ 1
```

**Comment by Midar** [comment]
First of all, building ObjFW with Clang from MSys2 does not work due to bugs in the version of Clang from MSys2. The multiple definitions errors you see while linking are a result of this. If you have both Clang and GCC installed, ObjFW will usually try to use Clang. To prevent it from doing that, specify OBJC=gcc to ./configure. The instructions from README-WINDOWS.md do not specify this as they only install GCC and not Clang and thus passing OBJC=gcc is not required.

That said, I just gave it a try with GCC on MSys2. I cannot reproduce the missing Sleep problem and both 32-bit and 64-bit builds work fine.

As for __MINGW32__ and __MINGW64__ both being defined, I think this is normal.

**Comment by IngwiePhoenix** [comment]
I configured with `OBJC=gcc` but now I always, with or without threads and runtime enabled, this:

```
arc.m:1:0: error: -fPIC ignored for target (all code is position independent) [-Werror]
 /*
 ^
cc1obj.exe: all warnings being treated as errors
Failed to compile arc.m (lib)!
```

**Comment by IngwiePhoenix** [comment]
Nice. So far so good. Now:

```
../../src/libobjfw.a(OFKernelEventObserver.o):OFKernelEventObserver.m:(.data+0x5d8): undefined reference to `__objc_class_name_OFKernelEventObserver_select'
collect2.exe: error: ld returned 1 exit status
Failed to link ofhash.exe!
../../buildsys.mk:213: die Regel für Ziel „ofhash.exe“ scheiterte
make[5]: *** [ofhash.exe] Fehler 1
../../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[4]: *** [all] Fehler 2
../buildsys.mk:129: die Regel für Ziel „ofhash“ scheiterte
make[3]: *** [ofhash] Fehler 2
../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[2]: *** [all] Fehler 2
buildsys.mk:129: die Regel für Ziel „utils“ scheiterte
make[1]: *** [utils] Fehler 2
buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2
```

This was configured with: `OBJC=gcc ./configure --disable-threads --disable-shared`

When I did another run but enabled threads:
```
In file included from property.m:27:0:
../threading.h: In function 'of_spinlock_lock':
../threading.h:215:3: error: implicit declaration of function 'Sleep' [-Werror=implicit-function-declaration]
   Sleep(0);
   ^
cc1obj.exe: all warnings being treated as errors
Failed to compile property.m!
../../buildsys.mk:482: die Regel für Ziel „property.o“ scheiterte
make[5]: *** [property.o] Fehler 1
../../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[4]: *** [all] Fehler 2
../buildsys.mk:129: die Regel für Ziel „runtime“ scheiterte
make[3]: *** [runtime] Fehler 2
../buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make[2]: *** [all] Fehler 2
buildsys.mk:129: die Regel für Ziel „src“ scheiterte
make[1]: *** [src] Fehler 2
buildsys.mk:116: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2
```

...so close to finally run ObjFW here. :) Sorry to bother you with all this but I have to do coding here at school and we are free to choose our stuff - and I'd like to use ObjFW.

**Comment by Midar** [comment]
> EDIT: Ahh. -fPIC is for shared libs. I disabled shared, it seems to compile now. If it doesnt run though ill let you know. :)

Well, there is definitely something wrong. For some reason, your system is not detected as being MinGW. The build system does not add -fPIC, yet it is added here. Please provide config.log for further debugging. This is absolutely a problem.

> ../../src/libobjfw.a(OFKernelEventObserver.o):OFKernelEventObserver.m:(.data+0x5d8): undefined reference to `__objc_class_name_OFKernelEventObserver_select'
> collect2.exe: error: ld returned 1 exit status

This means the .o file was built with Clang. So you did not do a clean build after using GCC.

> This was configured with: OBJC=gcc ./configure --disable-threads --disable-shared

Are you trying to work around the problem, or do you actually don't want a DLL and threads?

> In file included from property.m:27:0:
> ../threading.h: In function 'of_spinlock_lock':
> ../threading.h:215:3: error: implicit declaration of function 'Sleep' [-Werror=implicit-function-declaration]
>    Sleep(0);
>    ^

I don't get that. At this point, I'm quite certain your build environment is broken. Please provide config.log and output from gcc --version. Ideally, remove MSys2 entirely and then double-check that no other MinGW or GCC is installed on your system. After that, follow the instructions from README-WINDOWS.md.

**Comment by IngwiePhoenix** [comment]
I used `make distclean` and a release bundle and ran with three configurations, saved all the logs. Next step WOULD be to remove MSYS, but I need it a while longer and dont have time to reconfigure all the tools I use. So I'll just send the logs for now.

The convention I used in the filenames is pretty straight-forward...should be able to pick it up,

http://s000.tinyupload.com/index.php?file_id=55708370724739448110
(Hopefuly this doesn't spew too many ads :) I just wanted to upload a tiny TGZ file.)

**Comment by Midar** [comment]
> configure:2503: checking build system type
> configure:2517: result: x86_64-pc-msys
> configure:2537: checking host system type
> configure:2550: result: x86_64-pc-msys

Here it already goes wrong. This seems to come from the uname:

> uname -m = x86_64
> uname -r = 2.2.1(0.289/5/3)
> uname -s = MSYS_NT-6.3

I get that same uname when I start the MSys2 shortcut instead of the MinGW64/MinGW32 shortcut. However, when I use the MSys2 shortcut, the gcc command is not even in the PATH. However, in your case, it is, so there is definitely something very wrong with your build environment.

Please make sure you are using the correct shortcut (MinGW64 for 64 bit, MinGW32 for 32 bit) and that no other gcc is installed. If the issue persists, please do a clean reinstall and follow the instructions from README-WINDOWS.md carefully.

If it then works, I will clarify in the README-WINDOWS.md to be very careful to not use the MSys2 shortcut - though usually, that one should just complain that there's no compiler.

**Comment by IngwiePhoenix** [comment]
Copy that. I’ll do so. I added the /mingw64/bin path on my own - and never noticed multiple shortcuts. So, I’ll go ahead and try it out using the proper one then. Thanks for letting me know, i’ll tell you of my results when I get to it :)

**Comment by IngwiePhoenix** [comment]
Speak about a mishap. I used the mingw64 shortcut and BAM it worked! The host tripple showed `-mingw64` (or such) and it compiled perfectly fine with `OBJC=gcc`.

So - yes. People _have to_ use the `mingw64_shell.bat` in order to correctly compile ObjFW. I didn't try HEAD yet but 0.8-latest seems to work just fine.

Thank you SO MUCH for your assistance! I learned a lot from this issue - and you probably have one good example why people need to be warned about these shortcuts - or rather, launchers.
